### ISSUE TITLE
TF 2.0 save/restore DenseFeatures

### ISSUE THREAD

**Issue by batikus** [tensorflow/tensorflow#28923]
** Code **
Follwing this guide i created a model to classify structured data https://www.tensorflow.org/alpha/tutorials/keras/feature_columns :
```
feature_columns = []
headers = dataframe.columns.tolist()

# feature cols
for header in headers:
    temp = feature_column.numeric_column(header)
    feature_columns.append(feature_column.bucketized_column(temp, boundaries=[-89, -70, -65, -60, -55, -50, -40, -30, -20]))
    #feature_columns.append(temp) #old code used without buckets. only numeric columns


feature_layer = tf.keras.layers.DenseFeatures(feature_columns)

model = tf.keras.Sequential([
  feature_layer,
  layers.Dense(128, activation='relu'),
  layers.Dense(128, activation='relu'),
  layers.Dense(128, activation='relu'),
  layers.Dense(3, activation='sigmoid') #cambiare con il numero di zone
])
```

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Win10
- TensorFlow version (use command below): 2.0.0 alpha
- Python version: 3.6

**Describe the behavior**

Now when i try to save with the usual 
`model.save('path_to_my_model.h5')`
i cannot restore the model since i get an exception:

> Unknown layer: DenseFeatures

looking at some other [https://github.com/tensorflow/tensorflow/issues/26835](url) posts  I tried with 
`tf.keras.experimental.export_saved_model(model, 'experimental.h5')`
but here I get another exception:
> __init__() missing 1 required positional argument: 'feature_columns'

Tried to restore using 
```
for header in headers:
    temp = feature_column.numeric_column(header)
    feature_columns.append(feature_column.bucketized_column(temp, boundaries=[-89, -70, -65, -60, -55, -50, -40, -30, -20]))

feature_layer = tf.keras.layers.DenseFeatures(feature_columns)

new_model = tf.keras.models.load_model('guesser.h5',custom_objects={'DenseFeatures':feature_layer})

new_model.summary()
```
but it always give me the feature_columns error

**Comment by achandraa** [comment]
Tried with 2.0.0-alpha on Colab and was able to get the mentioned output.

**Comment by HubberDev** [comment]
I tried with tensorflow-2.0.0beta1, the issue is still there.

**Comment by HubberDev** [comment]
I spent time debugging issue in load_model(). Here is my theory:
- As for the error message "Unknown layer: DenseFeatures",  the problem happens with this line of code: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/serialization.py#L82.
DenseFeatures class is not part of the dict returned by the call to globals().

This problem can be solved (not sure if it is a correct way) by adding the following import to serialization.py:
from tensorflow.python.feature_column.feature_column_v2 import DenseFeatures

- After that, load_model() will complain about "init() missing 1 required positional argument: 'feature_columns'". This happens in the __init__ of DenseFeatures class. I think the root cause is that feature columns information is not serialized in the saved model file or is not read from the file correctly. From the deserialized model_config dict, I don't see feature columns.

Could anyone from the team take a look?  Thank you.

**Comment by HubberDev** [comment]
Digging more on this. based on this commit: https://github.com/tensorflow/tensorflow/commit/cd09510f9218220b872803c9ee70ba1dcd8c0f45#diff-99593c5667932dcaa61fb3e1b4921cc1, it seems like @karmel was fixing this issue and the fix hasn't made it to tf2.0 beta1 yet?

**Comment by karmel** [comment]
Can you try using the TensorFlow format for saving and restoring?

```
# feature cols
for header in headers:
    temp = feature_column.numeric_column(header)
    feature_columns.append(feature_column.bucketized_column(temp, boundaries=[-89, -70, -65, -60, -55, -50, -40, -30, -20]))
    #feature_columns.append(temp) #old code used without buckets. only numeric columns


feature_layer = tf.keras.layers.DenseFeatures(feature_columns)

model = tf.keras.Sequential([
  feature_layer,
  layers.Dense(128, activation='relu'),
  layers.Dense(128, activation='relu'),
  layers.Dense(128, activation='relu'),
  layers.Dense(3, activation='sigmoid') #cambiare con il numero di zone
])

model.save('output_path', format='tf')
```

That should work with the beta, though the change in question may have missed the cut. If it doesn't work, try with the nightly 2.0 release: https://pypi.org/project/tf-nightly-2.0-preview/

**Comment by cysmnl** [comment]
With tf.version 2.0.0-dev20190724, I tried saving a model `model.save('output_path', format='tf')` with no luck. I used the keras functional api to build the model.

**Comment by jvishnuvardhan** [comment]
@cysmnl Can you post a standalone code to reproduce the issue? Thanks!

**Comment by stefanondisponibile** [comment]
I am **very** confused about which API should I use: see [this simple Colab example](https://colab.research.google.com/drive/1CXESHGbE8pPZqtBPcsS5UV2tj_MYrxaH) with different approaches of *saving* and *restoring* a `tf.keras.Sequential` model with `feature_columns / DenseFeatures`.

---
**Save**:
```
🧐 Saved with keras APIs.
🧐 Saved with keras APIs in SavedModel format.
😵 Could **not** save with export_saved_model: __init__() missing 1 required positional argument: 'feature_columns'
🧐 Saved with tf.saved_model.save.
```
**Restore**:
```
😵 Couldn't load model via tf.keras.models.load_model (without custom objects): Unknown layer: DenseFeatures
😵 Couldn't load model via tf.keras.models.load_model (with custom objects): __init__() missing 1 required positional argument: 'feature_columns'
😵 Couldn't load model via tf.keras.experimental.load_from_saved_model: keras_model_tf/assets/saved_model.json; No such file or directory
😵 Couldn't load model via tf.keras.models.load_model: Passing a dictionary input to a Sequential Model which doesn't have FeatureLayer as the first layer is an error.
🧐 Model restored via `tf.saved.model.load('keras_model_tf')`
   Accuracy: 95.00%
🧐 Model restored via `tf.saved.model.load('tf_saved_model/1/')`
   Accuracy: 95.00%
```
---
I am now quite unsure about where the issue is and what are the best practices for exporting and restoring in this case.

**Comment by k-w-w** [comment]
I am currently submitting a fix for this issue, which will allow DenseFeatures to be saved and loaded using Keras APIs (`model.save` and `tf.keras.models.load_model`). 

The best practices can be summed up to:
 If you are planning to load the model back into Keras, use the Keras APIs for saving and loading. If you're not planning on using the Keras APIs to further train the model, then `tf.saved_model.save` and `tf.saved_model.load` is sufficient. 

I believe the docs need to be updated, as the experimental function `tf.keras.experimental.export_saved_model` is now deprecated.

**Comment by jvishnuvardhan** [comment]
@batikus Is this resolved? Please take a look at the [`feature_column`]((https://www.tensorflow.org/beta/tutorials/keras/feature_columns)) tutorial. I could save the model in "tf" and ".h5" formats without any issue.  Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/b6d63d35fd05058a3eb204891ca891c0/feature_columns.ipynb). If you still have an issue, please provide a simple standalone code to reproduce the issue. Thanks!

**Comment by jvishnuvardhan** [comment]
I am closing the issue as it was resolved in `tf-nightly-2.0-preview`. Please feel free to open it if the issue persists again. Thanks!

**Comment by EgorBEremeev** [comment]
> I am currently submitting a fix for this issue, which will allow DenseFeatures to be saved and loaded using Keras APIs (`model.save` and `tf.keras.models.load_model`).

@k-w-w , does this fix cover `tf.keras.experimental.SequenceFeatures` as well?

I save model by `tf.keras.models..save('MODEL_PATH.h5')`. Then while loading model by `tf.keras.models.load_model(MODEL_PATH.h5)`, I experience the similar error message as described in this issue: 

> Unknown layer: SequenceFeatures

I have not found separete issue for save\load problems concerning `SequenceFeatures` but my problem is reproduced in tensorflow **2.0.0-rc0** and **2.0.0b1** do not reproduced in **tf-nightly-2.0-preview**. It is not clear now open issue for `SequenceFeatures` or not.
